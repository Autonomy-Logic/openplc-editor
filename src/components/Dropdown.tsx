import { Menu, Transition } from '@headlessui/react'
import { FC, Fragment, PropsWithChildren, useEffect, useRef } from 'react'

import { classNames } from '@/utils'
/**
 * Defines the structure of a dropdown option.
 */
type DropdownOption = {
  label: string
  onClick: () => void
}
/**
 * Props for the Dropdown component.
 */
type DropdownProps = {
  show: boolean
  options: DropdownOption[]
  onClick?: () => void
  onAuxClick?: () => void
  onClose?: () => void
}
/**
 * Dropdown component that displays a list of options in a dropdown menu.
 */
const Dropdown: FC<PropsWithChildren<DropdownProps>> = ({
  children,
  options,
  show,
  onClick,
  onAuxClick,
  onClose,
}) => {
  const divRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    /**
     * Event handler for detecting clicks outside the dropdown.
     */
    const handleOutsideClick = (event: MouseEvent) => {
      if (
        show &&
        divRef.current &&
        !divRef.current.contains(event.target as Node)
      ) {
        onClose && onClose()
      }
    }
    /**
     * Add event listener to handle outside clicks
     */
    document.addEventListener('click', handleOutsideClick)
    /**
     * Remove event listener when component unmounts
     */
    return () => {
      document.removeEventListener('click', handleOutsideClick)
    }
  }, [onClose, show])

  return (
    <Menu
      as="div"
      className="relative z-10 inline-block text-left"
      ref={divRef}
    >
      <div>
        <Menu.Button onClick={onClick} onAuxClick={onAuxClick}>
          {children}
        </Menu.Button>
      </div>

      <Transition
        show={show}
        as={Fragment}
        enter="transition ease-out duration-100"
        enterFrom="transform opacity-0 scale-95"
        enterTo="transform opacity-100 scale-100"
        leave="transition ease-in duration-75"
        leaveFrom="transform opacity-100 scale-100"
        leaveTo="transform opacity-0 scale-95"
      >
        <Menu.Items
          static
          className="nodrag absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none"
        >
          <div className="py-1">
            {options.map(({ onClick, label }, index) => (
              <Menu.Item key={index}>
                {({ active }) => (
                  <button
                    onClick={() => {
                      onClick && onClick()
                      onClose && onClose()
                    }}
                    className={classNames(
                      'block w-full px-4 py-2 text-left text-sm transition-colors duration-300',
                      active
                        ? 'bg-open-plc-blue text-gray-50'
                        : 'text-gray-500 dark:text-gray-400',
                    )}
                  >
                    {label}
                  </button>
                )}
              </Menu.Item>
            ))}
          </div>
        </Menu.Items>
      </Transition>
    </Menu>
  )
}

export default Dropdown
