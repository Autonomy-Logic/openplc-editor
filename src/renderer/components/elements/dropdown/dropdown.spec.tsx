import '@testing-library/jest-dom/jest-globals'

import { fireEvent, getByTestId, render } from '@testing-library/react'

import Dropdown from '.'

// Tests for the Dropdown.Root component
describe('Dropdown Root Component', () => {
  // Test to check if the Root component renders correctly
  test('renders correctly', () => {
    // Render the Root component
    const { container } = render(<Dropdown.Root />)

    // Assert that the rendered component matches the snapshot
    expect(container).toMatchSnapshot()
  })

  // Test to check if HTMLAttributes are passed to the Root component
  test('passes HTMLAttributes to the dropdown component', () => {
    // Define a custom prop value
    const customProp = 'custom-prop-value'

    // Render the Root component with specific HTMLAttributes
    const { getByTestId } = render(<Dropdown.Root data-testid='root-test-id' data-custom-prop={customProp} />)

    // Get the Root element by test ID
    const rootElement = getByTestId('root-test-id')

    // Ensure rootElement is not null before accessing properties
    expect(rootElement).toBeTruthy()

    // Check if it's an element before calling getAttribute
    if (rootElement instanceof Element) {
      const rootProps = rootElement.getAttribute('data-custom-prop')
      expect(rootProps).toBe(customProp)
    } else {
      // If rootElement is not an element, fail the test
      fail('Root element is not an instance of Element')
    }
  })
})

// Tests for the Dropdown.Options component
describe('Options Component', () => {
  const options = [
    { label: 'Option 1', onClick: jest.fn() },
    { label: 'Option 2', onClick: jest.fn() },
    { label: 'Option 3', onClick: jest.fn() },
  ]

  // Test to check if the Options component renders correctly
  test('renders correctly', () => {
    const { container } = render(
      <Dropdown.Options
        setShowOptions={jest.fn()}
        options={options}
        setSelectedOption={jest.fn()}
        className='custom-class'
      />,
    )

    expect(container).toMatchSnapshot()
  })

  // Test to check if options are rendered correctly
  test('renders options correctly', () => {
    const { getByText } = render(
      <Dropdown.Options setShowOptions={jest.fn()} options={options} setSelectedOption={jest.fn()} />,
    )

    for (const option of options) {
      // Use the toBeInTheDocument matcher from @testing-library/jest-dom
      expect(getByText(option.label)).toBeInTheDocument()
    }
  })

  // Test to check if state is updated correctly when an option is clicked
  test('updates state and calls onClick when an option is clicked', () => {
    const setSelectedOption = jest.fn()
    const setShowOptions = jest.fn()

    const { getByText } = render(
      <Dropdown.Options setShowOptions={setShowOptions} options={options} setSelectedOption={setSelectedOption} />,
    )

    const optionToClick = options[1]
    fireEvent.click(getByText(optionToClick.label))

    expect(setSelectedOption).toHaveBeenCalledWith(optionToClick.label)
    expect(optionToClick.onClick).toHaveBeenCalled()
    expect(setShowOptions).toHaveBeenCalledWith(false)
  })
})

// Tests for the Dropdown.Select component
describe('Select Component', () => {
  const mockSetShowOptions = jest.fn()

  // Test to check if the Select component renders correctly
  test('renders correctly', () => {
    const { container } = render(
      <Dropdown.Select
        selectedOption='Option 1'
        Icon={undefined}
        setShowOptions={mockSetShowOptions}
        showOptions={false}
        placeholder='Select an option'
      />,
    )

    expect(container).toMatchSnapshot()
  })

  // Test to check if placeholder text is displayed correctly
  test('displays placeholder text', () => {
    const { getByText } = render(
      <Dropdown.Select
        selectedOption='Option 1'
        Icon={undefined}
        setShowOptions={mockSetShowOptions}
        showOptions={false}
        placeholder='Select an option'
      />,
    )

    expect(getByText('Select an option')).toBeInTheDocument()
  })

  // Test to check if selected option text is displayed correctly
  test('displays selected option text', () => {
    const { getByText } = render(
      <Dropdown.Select
        selectedOption='Option 1'
        Icon={undefined}
        setShowOptions={mockSetShowOptions}
        showOptions={false}
        placeholder='Select an option'
      />,
    )

    expect(getByText('Option 1')).toBeInTheDocument()
  })

  // Test to check if icon is displayed correctly
  test('displays icon', () => {
    const { getByTestId } = render(
      <Dropdown.Select
        selectedOption='Option 1'
        Icon={undefined}
        setShowOptions={mockSetShowOptions}
        showOptions={false}
        placeholder='Select an option'
      />,
    )

    expect(getByTestId('dropdown-icon')).toBeInTheDocument()
  })

  // Test to check if setShowOptions is called on button click
  test('calls setShowOptions on button click', () => {
    const { getByText } = render(
      <Dropdown.Select
        selectedOption='Option 1'
        Icon={undefined}
        setShowOptions={mockSetShowOptions}
        showOptions={false}
        placeholder='Select an option'
      />,
    )

    fireEvent.click(getByText('Select an option'))
    expect(mockSetShowOptions).toHaveBeenCalledTimes(1)
  })
})
