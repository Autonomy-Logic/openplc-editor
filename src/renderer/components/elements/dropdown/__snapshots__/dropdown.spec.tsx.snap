// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Root Component renders correctly 1`] = `
<div>
  <div
    class="relative h-14"
  />
</div>
`;

exports[`Options Component renders correctly 1`] = `
<div>
  <div
    class="custom-class"
  >
    <div
      class="w-full h-full flex flex-col"
    >
      <button
        class="cursor-pointer hover:bg-neutral-50"
        type="button"
      >
        <p
          class="flex-1 text-neutral-1000 text-left py-[10px] pl-4"
        >
          Option 1
        </p>
      </button>
      <button
        class="cursor-pointer hover:bg-neutral-50"
        type="button"
      >
        <p
          class="flex-1 text-neutral-1000 text-left py-[10px] pl-4"
        >
          Option 2
        </p>
      </button>
      <button
        class="cursor-pointer hover:bg-neutral-50"
        type="button"
      >
        <p
          class="flex-1 text-neutral-1000 text-left py-[10px] pl-4"
        >
          Option 3
        </p>
      </button>
    </div>
  </div>
</div>
`;

exports[`Select Component renders correctly 1`] = `
<div>
  <button
    class="select-none"
    type="button"
  >
    Select an option
    <p
      class="flex justify-between w-28"
    >
      <span
        class="text-black "
      >
        Option 1
      </span>
      <span
        class=" rotate-0 transition-all "
      >
        <span>
          Icon
        </span>
      </span>
    </p>
  </button>
</div>
`;
